zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/migration/VersionCompatibilityCheckTest.java

```diff
@@ -0,0 +1,115 @@
+      assertThat(check("8.0.0-alpha1", "8.0.0-beta1"))
+          .isInstanceOf(Incompatible.ChangeOfPreReleaseVersion.class);
+      assertThat(check("8.0.0-beta1", "8.0.0-rc1"))
+          .isInstanceOf(Incompatible.ChangeOfPreReleaseVersion.class);
+      assertThat(check("8.0.0-rc1", "8.0.0"))
+          .isInstanceOf(Incompatible.ChangeOfPreReleaseVersion.class);
+      assertThat(check("8.0.0-alpha1", "8.1.0"))
+          .isInstanceOf(Incompatible.ChangeOfPreReleaseVersion.class);
+      assertThat(check("8.0.0-alpha2", "8.0.1"))
+          .isInstanceOf(Incompatible.ChangeOfPreReleaseVersion.class);
+    }
+
+    @Test
+    public void shouldRejectMajorUpgrades() {
+      assertThat(check("8.0.0", "9.0.0")).isInstanceOf(Incompatible.MajorUpgrade.class);
+      assertThat(check("8.1.0", "9.0.0")).isInstanceOf(Incompatible.MajorUpgrade.class);
+    }
+
+    @Test
+    public void shouldRejectSkippedMinorVersions() {
+      assertThat(check("8.0.0", "8.2.0")).isInstanceOf(Incompatible.SkippedMinorVersion.class);
+      assertThat(check("8.0.2", "8.3.0")).isInstanceOf(Incompatible.SkippedMinorVersion.class);
+    }
+  }
+
+  @Nested
+  final class IndeterminateResults {
+    @Test
+    void shouldDetectUnknownVersions() {
+      assertThat(check(null, "8.0.0")).isInstanceOf(Indeterminate.PreviousVersionUnknown.class);
+      assertThat(check("8.0.0", null)).isInstanceOf(Indeterminate.CurrentVersionUnknown.class);
+    }
+
+    @Test
+    void shouldDetectInvalidVersions() {
+      assertThat(check("dev", "1.2.3")).isInstanceOf(Indeterminate.PreviousVersionInvalid.class);
+      assertThat(check("1.2.3", "dev")).isInstanceOf(Indeterminate.CurrentVersionInvalid.class);
+    }
+  }
+
+  @Nested
+  final class CompatibleResults {
+
+    @Test
+    void shouldAcceptPatchUpgrades() {
+      assertThat(check("8.0.0", "8.0.1")).isInstanceOf(Compatible.PatchUpgrade.class);
+      assertThat(check("8.1.4", "8.1.5")).isInstanceOf(Compatible.PatchUpgrade.class);
+    }
+
+    @Test
+    void shouldAcceptMinorUpgrades() {
+      assertThat(check("8.0.0", "8.1.0")).isInstanceOf(Compatible.MinorUpgrade.class);
+      assertThat(check("8.1.3", "8.2.0")).isInstanceOf(Compatible.MinorUpgrade.class);
+    }
+
+    @Test
+    void shouldAcceptSameVersions() {
+      assertThat(check("8.0.0", "8.0.0")).isInstanceOf(Compatible.SameVersion.class);
+      assertThat(check("8.0.0-alpha1", "8.0.0-alpha1")).isInstanceOf(Compatible.SameVersion.class);
+      assertThat(check("8.1.0-alpha1+build123", "8.1.0-alpha1+build234"))
+          .isInstanceOf(Compatible.SameVersion.class);
+    }
+
+    @Test
+    void shouldAcceptUpgradeToPreReleaseVersions() {
+      assertThat(check("8.0.0", "8.0.1-alpha1")).isInstanceOf(Compatible.PatchUpgrade.class);
+      assertThat(check("8.0.0", "8.1.0-beta1")).isInstanceOf(Compatible.MinorUpgrade.class);
+      assertThat(check("8.1.2", "8.2.3-rc1")).isInstanceOf(Compatible.MinorUpgrade.class);
+    }
```

Comment:
❌ This would be dangerous for production clusters. After any of these updates, they can no longer update to a regular release. We should reject these.

Reply:
So this would extend the checks to not only forbid going away from a pre-release version but also prevent switching to a pre-release version.
I feel like both checks severely limit the usefulness of our alpha releases. For example, they make it impossible for customers to accept some risk and start using an alpha version to test out a new feature on their existing cluster.

Reply:
As discussed, I'll prevent any version change **to** or **from** a pre-release version.

Reply:
Thanks @oleschoenburg for taking care of this 🙇